File: trunk/Event/eventsupport/test/testeventsupport/java/src/cdmw/event/test/ClientThread.java
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI. It is derivative work based on PERCO Copyright (C) THALES
 * 2000-2003. All rights reserved.
 * 
 * Copyright (C) THALES 2004-2005. All rights reserved
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with CARDAMOM; see the file COPYING. If not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


package cdmw.event.test;

import cdmw.testutils.Testable;
import cdmw.common.Locations;
import cdmw.common.Options;

import cdmw.ossupport.OS;
import cdmw.ossupport.ThreadedProcess;
import cdmw.orbsupport.ORBUtils;

import cdmw.tools.ProcessAdmin;
import cdmw.commonsvcs.naming.NamingInterface;
import cdmw.namingandrepository.RepositoryInterface;

import com.thalesgroup.CdmwPlatformMngt.ProcessDelegate;
import com.thalesgroup.CdmwPlatformMngt.ProcessDelegateHelper;

public class ClientThread extends Thread {

    private org.omg.CORBA.ORB orb;

    private String callbackIOR;

    private String nameServicePort;

    private String managerArgs;

    private static final String profileManagerService
        = "cdmw.eventadmin.EventChannelManager";

    private int status = MainTest.SUCCESS;
    
    private int timescale = 1;

    public ClientThread(org.omg.CORBA.ORB orb, String nameServicePort,
        String callbackIOR) {

        this.orb = orb;
        this.nameServicePort = nameServicePort;
          this.callbackIOR = callbackIOR;
        this.managerArgs =
            " -ORBconfig ./obtest.conf " +
            Options.CALLBACK_IOR_OPTION + "=" + callbackIOR + " " +
            Options.PROCESS_PORT_OPTION + "=15327 " +
            Options.LOCALISATION_SERVICE_OPTION + "=15328" +
            " --ProfileXMLFile=./EventChannelConf1.xml " +
            Options.PROCESS_XML_FILE_OPTION + "=./Server.xml " +
            Options.VALIDATE_XML_OPTION;
        this.timescale = Testable.getTimescale();

    }

    public int getStatus() {
        return this.status;
    }

    public void run() {

        try {
            // init main server of this test: PROC_000
            String procUrl = "corbaname::localhost:" + nameServicePort
                + "#CDMW.I/ProcessCallback.simulated/CDMW Test Application/PROC_000";
            ProcessDelegate proc = ProcessDelegateHelper.narrow(orb.string_to_object(procUrl));
            cdmw.tools.ProcessAdmin admin = new ProcessAdmin(orb, proc);

            // send initialise order
            String commands = "init LAST 2\n exit \n";
            admin.run(new java.io.BufferedReader(
                new java.io.StringReader(commands)), System.out);
            OS.sleep(timescale * 5000);

            // send run order
            commands = "run\n exit \n";
            admin.run(new java.io.BufferedReader(
                new java.io.StringReader(commands)), System.out);
            OS.sleep(timescale * 3000);

            //
            // Get NamingInterface object from RepositoryInterface
            //
            NamingInterface ni = null;
            try {
                ni = RepositoryInterface.getDomainNamingInterface(
                    Locations.CDMW_SERVICES_NAME_DOMAIN);
            } catch(com.thalesgroup.CdmwNamingAndRepository.NoNameDomain nnd) {
                nnd.printStackTrace();
            } catch(com.thalesgroup.CdmwNamingAndRepository.InvalidName in) {
                in.printStackTrace();
            }

            // I) TEST UNTYPED EVENT CHANNEL FACTORY
            testUntypedFactory(ni);

            OS.sleep(timescale * 5000);

            // II) TEST TYPED EVENT CHANNEL FACTORY
            testTypedFactory(ni);

        } catch(Exception e) {
            e.printStackTrace();
            status = MainTest.FAILURE;
        }

    }

    public void testUntypedFactory(NamingInterface ni)
        throws Exception {

        //
        // Get a reference on the EventChannelFactory
        //
        org.omg.CORBA.Object obj
            = ni.resolve("EVENT/FACTORIES/DefaultEventChannelFactory");
        com.thalesgroup.CdmwEvent.EventChannelFactory factory
            = com.thalesgroup.CdmwEvent.EventChannelFactoryHelper.narrow(obj);

        if (factory==null) {
            System.out.println("defaultEventChannelFactory is a nil object reference");
            return;
        }

        // Testing TestEventChannelFactory
        EventChannelFactoryTest factoryTest
            = new EventChannelFactoryTest(factory);
        factoryTest.getHeader();
        factoryTest.startTests();

        // Starting the EventChannelManager
        String url = "corbaname::localhost:" + nameServicePort
            + "#CDMW.I/ProcessCallback.simulated/CDMW Test Application/PROC_001";
        startEventChannelManager(url);

        // Testing TestEventChannelFactory continue
        factoryTest.nextTest();

        // Starting Eventservice
        ThreadedProcess eventServer
            = ORBUtils.startEventService("-ORBconfig ./obtest.conf");
        OS.sleep(timescale * 10000);
        System.out.println("Event service started");

        // Testing TestEventChannelFactory ending
        factoryTest.nextTest();

        if (orb!=null) {
            try {
                //OS.killProcess(profileManager);
                stopEventChannelManager(url);
                OS.killProcess(eventServer);
            } catch(Exception e) {
                e.printStackTrace();
            }
        }

        factoryTest.getSummary();

    }


    public void testTypedFactory(NamingInterface ni)
        throws Exception {

        //
        // Get a reference on the TypedEventChannelFactory
        //
        org.omg.CORBA.Object obj
            = ni.resolve("EVENT/FACTORIES/DefaultTypedEventChannelFactory");
        com.thalesgroup.CdmwEvent.TypedEventChannelFactory typedFactory
            = com.thalesgroup.CdmwEvent.TypedEventChannelFactoryHelper.narrow(obj);

        if (typedFactory == null) {
            System.out.println("defaultTypedEventChannelFactory is a nil object reference");
            return;
        }

        // Testing TypedEventChannelFactory
        TypedEventChannelFactoryTest typedFactoryTest
            = new TypedEventChannelFactoryTest(typedFactory);
        typedFactoryTest.getHeader();
        typedFactoryTest.startTests();

        // Starting the EventChannelManager
        String url = "corbaname::localhost:" + nameServicePort
            + "#CDMW.I/ProcessCallback.simulated/CDMW Test Application/PROC_002";
        startEventChannelManager(url);

        // Testing TypedEventChannelFactory continue
        typedFactoryTest.nextTest();

        // Starting typed Eventservice
        // Launch irserv needed by eventserv for typed event
        //ThreadedProcess irServer
        //    = OS.createProcess("irserv -ORBconfig ./obtest.conf -I../../../../../export/idl ../../idl/Echo.idl");
        ThreadedProcess eventServer
            = ORBUtils.startEventService("-t -ORBconfig ./obtest.conf");
        OS.sleep(timescale * 10000);
        System.out.println("Event service started");

        // Testing TestEventChannelFactory ending
        typedFactoryTest.nextTest();

        if (orb!=null) {
            try {
                //OS.killProcess(profileManager);
                stopEventChannelManager(url);
                OS.killProcess(eventServer);
                //OS.killProcess(irServer);
            } catch(Exception ex) {
                ex.printStackTrace();
            }
        }

        typedFactoryTest.getSummary();

    }

    public void startEventChannelManager(String url) throws Exception {

        ThreadedProcess profileManager = OS.createJavaProcess(
            profileManagerService + managerArgs, Testable.getProperties());
        System.out.println("Launch EventChannelManager " + managerArgs);
        OS.sleep(timescale * 10000);
        System.out.println("EventChannelManager started");

        ProcessDelegate proc = ProcessDelegateHelper.narrow(orb.string_to_object(url));
        ProcessAdmin admin = new ProcessAdmin(orb, proc);

        // send initialise order
        String commands = "init LAST 2\n exit \n";
        admin.run(new java.io.BufferedReader(
            new java.io.StringReader(commands)), System.out);
        OS.sleep(timescale * 5000);

        // send run order
        commands = "run\n exit \n";
        admin.run(new java.io.BufferedReader(
            new java.io.StringReader(commands)), System.out);
        OS.sleep(timescale * 3000);

    }

    public void stopEventChannelManager(String url) throws Exception {

        ProcessDelegate proc = ProcessDelegateHelper.narrow(orb.string_to_object(url));
        ProcessAdmin admin = new ProcessAdmin(orb, proc);

        // send stop order
        String commands = "stop\n exit \n";
        admin.run(new java.io.BufferedReader(
            new java.io.StringReader(commands)), System.out);

    }

}

File: tags/CARDAMOM_V2_1_CE/Event/eventsupport/test/testeventsupport/c++/src/ClientThread.cpp
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI. It is derivative work based on PERCO Copyright (C) THALES
 * 2000-2003. All rights reserved.
 * 
 * Copyright (C) THALES 2004-2005. All rights reserved
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with CARDAMOM; see the file COPYING. If not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


#include <string>
#include <sstream>
#include <cctype>
#include <algorithm>

#include "testeventsupport/ClientThread.hpp"

#include "Repository/repositoryinterface/RepositoryInterface.hpp"
#include "Foundation/commonsvcs/naming/NamingInterface.hpp"
#include "Foundation/commonsvcs/naming/NamingUtil.hpp"

#include "Foundation/common/System.hpp"
#include "Foundation/common/Locations.hpp"
#include "Foundation/common/Options.hpp"
#include "Foundation/ossupport/OS.hpp"
#include "Foundation/orbsupport/OrbSupport.hpp"
#include "Foundation/common/Exception.hpp"
#include "Foundation/common/Assert.hpp"
#include "Foundation/testutils/Testable.hpp"

#include "SystemMngt/tools/ProcessAdmin.hpp"

#include "testeventsupport/TestEventChannelFactory.hpp"
#include "testeventsupport/TestTypedEventChannelFactory.hpp"

namespace
{
}; // End anonymous namespace

namespace Cdmw
{
namespace Event
{

//----------------------------------------------------------------------------
ClientThread::ClientThread(CORBA::ORB_ptr orb, const char* nameServicePort, const char* callbackIOR,
                           const char* cdmwIdlPath) 
    throw (CORBA::SystemException)
    : m_orb(CORBA::ORB::_duplicate(orb)),
      m_callbackIOR(callbackIOR),
      m_nameServicePort(nameServicePort),
      m_cdmwIdlPath(cdmwIdlPath)
{
}

//----------------------------------------------------------------------------
ClientThread::~ClientThread() throw()
{
}

//----------------------------------------------------------------------------
void ClientThread::run() throw()
{
    using namespace Cdmw::Common;
    using namespace CdmwEvent;
    using namespace Cdmw::OsSupport;

    // init main server of this test: PROC_000
    CdmwPlatformMngt::ProcessDelegate_var proc = CdmwPlatformMngt::ProcessDelegate::_nil();
    std::string proc_url("corbaname::localhost:");
    proc_url += m_nameServicePort;
    proc_url += "#CDMW.I/ProcessCallback.simulated/CDMW Test Application/PROC_000";
    CORBA::Object_var obj = m_orb->string_to_object(proc_url.c_str());
    
    if (!CORBA::is_nil(obj.in())) 
    {
        proc = CdmwPlatformMngt::ProcessDelegate::_narrow(obj.in());
    }
    else
    {
        std::cerr << " **** TEST FAILED AT LINE " << __LINE__ 
            << " IN FILE: " << __FILE__ << std::endl;
        return;
    }
    
    Cdmw::Tools::ProcessAdmin admin(m_orb.in(),proc.in());
    
    int timescale = Cdmw::TestUtils::Testable::get_timescale();

    // send initialise order
    std::string commands("init LAST 2\n exit \n");
    std::istringstream commands_is(commands);
    admin.run(commands_is,std::cout);
    OS::sleep(timescale*5000);
    
    // send run order
    commands = "run\n exit \n";
    commands_is.str(commands);
    admin.run(commands_is,std::cout);
    OS::sleep(timescale*3000);
  
    //
    // Get NamingInterface object from RepositoryInterface
    //
    Cdmw::CommonSvcs::Naming::NamingInterface ni
        = Cdmw::NamingAndRepository::RepositoryInterface::get_domain_naming_interface(Locations::CDMW_SERVICES_NAME_DOMAIN);
    
    // I) TEST UNTYPED EVENT CHANNEL FACTORY
    //
    //Get a reference on the EventChannelFactory
    //
    obj = ni.resolve("EVENT/FACTORIES/DefaultEventChannelFactory");
    EventChannelFactory_var factory = EventChannelFactory::_narrow(obj.in());
    
    if (CORBA::is_nil(factory.in())) 
    {
        std::cerr << "defaultEventChannelFactory is a nil object reference" << std::endl;
        std::cerr << " **** TEST FAILED AT LINE " << __LINE__ 
                  << " IN FILE: " << __FILE__ << std::endl;
        return;
    }

    // Testing TestEventChannelFactory 
    TestEventChannelFactory testFactory(factory.in());
    testFactory.start();
    
    // Starting the EventChannelManager
    std::string profileManagerService="cdmw_event_channel_manager";
#if (CDMW_ORB_VDR == orbacus )
    std::string managerArgs=" -ORBconfig obtest.conf ";        
#else
    std::string managerArgs="";        
#endif
    managerArgs += Options::CALLBACK_IOR_OPTION;
    managerArgs += "=" + m_callbackIOR + " " 
                + Options::PROCESS_PORT_OPTION + "=15327 " 
                + Options::LOCALISATION_SERVICE_OPTION+ "=15328 " 
                + " --ProfileXMLFile=EventChannelConf1.xml " 
                + Options::PROCESS_XML_FILE_OPTION + "=Server.xml "
                    + Options::VALIDATE_XML_OPTION;
    std::string dummy;
    
    std::cout << " launch EventChannelManager " << managerArgs << std::endl;
    OS::ProcessId idProfileManager;
    idProfileManager = OS::create_process(profileManagerService, managerArgs);
    //std::cin >> dummy;
     
    OS::sleep(timescale*10000);
    std::cout << "EventChannelManager started..." << std::endl;
    
    // EventChannelManager Proc URL is CDMW Test Application/PROC_001 while main 
    // of this test is PROC_000
    std::cout << " Getting eventchannelmanager ProcessCallback" << std::endl;
    CdmwPlatformMngt::ProcessDelegate_var proc_ec_mng = CdmwPlatformMngt::ProcessDelegate::_nil();
    proc_url = "corbaname::localhost:" + m_nameServicePort;
    proc_url += "#CDMW.I/ProcessCallback.simulated/CDMW Test Application/PROC_001";
    obj = m_orb->string_to_object(proc_url.c_str());
    
    if (!CORBA::is_nil(obj.in())) 
    {
       proc_ec_mng  = CdmwPlatformMngt::ProcessDelegate::_narrow(obj.in());
    }
    else
    {
        std::cerr << " **** TEST FAILED AT LINE " << __LINE__ 
            << " IN FILE: " << __FILE__ << std::endl;
        return;
    }
    
    std::cout << " Getting Process admin" << std::endl;
    Cdmw::Tools::ProcessAdmin admin1(m_orb.in(),proc_ec_mng.in());
    
    // send initialise order
    commands = "init LAST 2\n exit \n";
    commands_is.str(commands);
    admin1.run(commands_is,std::cout);
    OS::sleep(timescale*5000);
    
    // send run order
    commands = "run\n exit \n";
    commands_is.str(commands);
    admin1.run(commands_is,std::cout);
    OS::sleep(timescale*3000);
        
    // Testing TestEventChannelFactory continue
    testFactory.next_test();

#if (CDMW_ORB_VDR == orbacus )
    // Starting Eventservice
    OS::ProcessId idEventServer;
    idEventServer = OS::create_process("eventserv", "-ORBconfig obtest.conf");
    OS::sleep(timescale*10000);
    std::cout << "Eventserv started" << std::endl;
#endif

    // Testing TestEventChannelFactory ending
    testFactory.next_test();

    if (!CORBA::is_nil(m_orb.in()))
    {
        try
        {
            // send stop order
            commands = "stop\n exit \n";
            commands_is.str(commands);
            std::cout << "Stop cdmw_event_channel_manager" << std::endl;
            admin1.run(commands_is,std::cout);
            
#if (CDMW_ORB_VDR == orbacus )
            OS::kill_process(idEventServer);
#endif
        }
        catch(const CORBA::Exception& ex)
        {
            std::cerr << ex << std::endl;
            std::cerr << " **** TEST FAILED AT LINE " << __LINE__ 
                      << " IN FILE: " << __FILE__ << std::endl;
        }
    }   
     OS::sleep(timescale*5000);

#if (CDMW_ORB_VDR == orbacus )
    // II) TEST TYPED EVENT CHANNEL FACTORY
    //
    // Get a reference on the TypedEventChannelFactory
    //
    obj = ni.resolve("EVENT/FACTORIES/DefaultTypedEventChannelFactory");
    TypedEventChannelFactory_var typedFactory = TypedEventChannelFactory::_narrow(obj.in());
    
    if (CORBA::is_nil(typedFactory.in())) 
    {
        std::cerr << "defaultTypedEventChannelFactory is a nil object reference" << std::endl;
        std::cerr << " **** TEST FAILED AT LINE " << __LINE__ 
                  << " IN FILE: " << __FILE__ << std::endl;
        return;
    }

    // Testing TestEventChannelFactory 
    TestTypedEventChannelFactory testTypedFactory(typedFactory.in());
    testTypedFactory.start();
    
    // Starting the EventChannelManager
    idProfileManager
        = OS::create_process(profileManagerService, managerArgs);
    std::cout << " launch EventChannelManager " << managerArgs;
    //std::cin >> dummy;
    OS::sleep(timescale*10000);
    std::cout << "EventChannelManager started" << std::endl;

    proc = CdmwPlatformMngt::ProcessDelegate::_nil();
    // EventChannelManager Proc URL shall be incremented (stoppped, then restarted)
    // and is now CDMW Test Application/PROC_002
    proc_url = "corbaname::localhost:" + m_nameServicePort;
    proc_url += "#CDMW.I/ProcessCallback.simulated/CDMW Test Application/PROC_002";
    obj = m_orb->string_to_object(proc_url.c_str());
    
    if (!CORBA::is_nil(obj.in())) 
    {
        proc = CdmwPlatformMngt::ProcessDelegate::_narrow(obj.in());
    }
    else
    {
        std::cerr << " **** TEST FAILED AT LINE " << __LINE__ 
            << " IN FILE: " << __FILE__ << std::endl;
        return;
    }
    
    Cdmw::Tools::ProcessAdmin admin2(m_orb.in(),proc.in());
    
    // send initialise order
    commands = "init LAST 2\n exit \n";
    commands_is.str(commands);
    admin2.run(commands_is,std::cout);
    OS::sleep(timescale*5000);
    
    // send run order
    commands = "run\n exit \n";
    commands_is.str(commands);
    admin2.run(commands_is,std::cout);
          
    // Testing TestTypedEventChannelFactory continue
    testTypedFactory.next_test();

    // Starting typed Eventservice
    // Launch irserv needed by eventserv for typed event
    std::string irservArgs = "-ORBconfig obtest.conf -I";        
    irservArgs += m_cdmwIdlPath;
    irservArgs += " Echo.idl";
     
    OS::ProcessId idIRServer = OS::create_process("irserv", irservArgs);
    
    idEventServer = OS::create_process("eventserv", "-t -ORBconfig obtest.conf ");
    
    OS::sleep(timescale*10000);
    std::cout << "Eventserv started" << std::endl;


    // Testing TestEventChannelFactory ending
    testTypedFactory.next_test();

    if (!CORBA::is_nil(m_orb.in()))
    {
        try
        {
            // send stop order
            commands = "stop\n exit \n";
            commands_is.str(commands);
            std::cout << "Stop cdmw_event_channel_manager" << std::endl;
            admin2.run(commands_is,std::cout);

            OS::kill_process(idEventServer);
            OS::kill_process(idIRServer);
            
            m_orb -> shutdown(false);
        }
        catch(const CORBA::Exception& ex)
        {
            std::cerr << ex << std::endl;
            std::cerr << " **** TEST FAILED AT LINE " << __LINE__ 
                      << " IN FILE: " << __FILE__ << std::endl;
        }
    }   
#else
     if (!CORBA::is_nil(m_orb.in()))
    {
        try
        {
            std::cout << "shutdown called" << std::endl;
            m_orb -> shutdown(false);
        }
        catch(const CORBA::Exception& ex)
        {
            std::cerr << ex << std::endl;
            std::cerr << " **** TEST FAILED AT LINE " << __LINE__ 
                      << " IN FILE: " << __FILE__ << std::endl;
        }
    }   

#endif
}

}; // namespace Event
}; // namespace Cdmw


